File: app/src/main/java/com/example/dictionaryapp/DictEntryAdapter.java
package com.example.dictionaryapp;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.TextView;

import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.example.dictionaryapp.database.DBHelper;
import com.example.dictionaryapp.database.DictEntry;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.IOException;
import java.text.DecimalFormat;
import java.util.List;

public class DictEntryAdapter extends RecyclerView.Adapter<DictEntryAdapter.ViewHolder> {

    private List<DictEntry> mData;
    private Context mContext;
    private int lastPosition = -1;


    // data is passed into the constructor
    public DictEntryAdapter(Context context, List<DictEntry> data) {
        this.mContext = context;
        this.mData = data;
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_listentry, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final DictEntry entry = mData.get(position);
        holder.titleView.setText(entry.getKanji());

        holder.secondaryView.setText(entry.getReading());
        String meaning = entry.getMeaning();
        if (meaning.length() >= 100) {
            meaning = meaning.substring(0, 100) + "...";
        }
        holder.detailView.setText(meaning);
        holder.dictName.setText(entry.getShortenedDictName());

        setAnimation(holder.itemView, position);

        holder.card.setOnClickListener(v -> {
            System.out.println("HI");
            Intent intent = new Intent(mContext, WordDetailActivity.class);
            intent.putExtra("word", entry.getKanji());
            intent.putExtra("reading", entry.getReading());
            intent.putExtra("meaning", entry.getMeaning());
            intent.putExtra("pitch", entry.getPitchAccent());
            Integer frequency = entry.getFreq();
            if (frequency != null) {
                DecimalFormat formatter = new DecimalFormat("#,###");
                intent.putExtra("freq", "Freq: " + formatter.format(entry.getFreq()));
            } else {
                intent.putExtra("freq", "");
            }

            JSONArray splittedTags = new JSONArray();
            try {
                splittedTags = DBHelper.getSplittedTags(entry, mContext);
            } catch (IOException e) {
                e.printStackTrace();
            } catch (JSONException e) {
                e.printStackTrace();
            }
            intent.putExtra("tags", splittedTags.toString());

            mContext.startActivity(intent);
        });


    }

    private void setAnimation(View viewToAnimate, int position) {
        // If the bound view wasn't previously displayed on screen, it's animated
        if (position > lastPosition) {
            Animation animation = AnimationUtils.loadAnimation(mContext, android.R.anim.fade_in);
            viewToAnimate.startAnimation(animation);
            lastPosition = position;
        }
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return mData.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView titleView;
        TextView secondaryView;
        TextView detailView;
        TextView dictName;
        CardView card;

        ViewHolder(View itemView) {
            super(itemView);
            card = itemView.findViewById(R.id.card);
            titleView = itemView.findViewById(R.id.wordTextView);
            secondaryView = itemView.findViewById(R.id.secondaryTextCard);
            detailView = itemView.findViewById(R.id.meaningTextView);
            dictName = itemView.findViewById(R.id.dictName);
        }

        @Override
        public void onClick(View v) {

        }
    }

    // convenience method for getting data at click position
    DictEntry getItem(int id) {
        return mData.get(id);
    }

}

File: app/src/main/java/com/kamui/rin/DictEntryAdapter.kt
package com.kamui.rin

import android.content.Context
import android.content.Intent
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.kamui.rin.db.DictEntry
import com.kamui.rin.db.getTagsFromSplitted
import java.text.DecimalFormat

class DictEntryAdapter(private val mContext: Context, data: List<DictEntry>) :
    RecyclerView.Adapter<DictEntryAdapter.ViewHolder>() {
    private val mData: List<DictEntry> = data
    private var lastPosition = -1

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view: View =
            LayoutInflater.from(parent.context).inflate(R.layout.layout_listentry, parent, false)
        return ViewHolder(view)
    }

    @RequiresApi(Build.VERSION_CODES.KITKAT)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val entry = mData[position]
        holder.titleView.text = entry.kanji
        holder.secondaryView.text = entry.reading
        var meaning: String = entry.getMeaning()
        if (meaning.length >= 100) {
            meaning = meaning.substring(0, 100) + "..."
        }

        holder.detailView.text = meaning
        holder.dictName.text = entry.shortenedDictName
        setAnimation(holder.itemView, position)

        holder.card.setOnClickListener {
            val intent = Intent(mContext, WordDetailActivity::class.java)
            intent.putExtra("word", entry.kanji)
            intent.putExtra("reading", entry.reading)
            intent.putExtra("meaning", entry.getMeaning())
            intent.putExtra("pitch", entry.pitchAccent)

            val frequency: Int? = entry.freq
            val formatter = DecimalFormat("#,###")
            intent.putExtra("freq", "Freq: " + formatter.format(frequency))

            val splittedTags = getTagsFromSplitted(entry, mContext)
            intent.putExtra("tags", splittedTags.toTypedArray())
            mContext.startActivity(intent)
        }
    }

    private fun setAnimation(viewToAnimate: View, position: Int) {
        // If the bound view wasn't previously displayed on screen, it's animated
        if (position > lastPosition) {
            val animation = AnimationUtils.loadAnimation(mContext, android.R.anim.fade_in)
            viewToAnimate.startAnimation(animation)
            lastPosition = position
        }
    }

    // total number of rows
    override fun getItemCount(): Int {
        return mData.size
    }

    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        var titleView: TextView = itemView.findViewById(R.id.wordTextView)
        var secondaryView: TextView = itemView.findViewById(R.id.secondaryTextCard)
        var detailView: TextView = itemView.findViewById(R.id.meaningTextView)
        var dictName: TextView = itemView.findViewById(R.id.dictName)
        var card: CardView = itemView.findViewById(R.id.card)
        override fun onClick(v: View) {}
    }
}